pkg/runtime/pubsub/default_bulkpub.go

```diff
@@ -0,0 +1,132 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package pubsub
+
+import (
+	"context"
+	"errors"
+
+	"golang.org/x/sync/errgroup"
+
+	contribPubsub "github.com/dapr/components-contrib/pubsub"
+	"github.com/dapr/dapr/utils"
+)
+
+const (
+	bulkPublishSeriallyKey       string = "bulkPublishSerially"
+	bulkPublishMaxConcurrencyKey string = "bulkPublishMaxConcurrency"
+
+	defaultBulkPublishMaxConcurrency int = 100
+)
+
+var ErrBulkPublishFailure = errors.New("bulk publish failed")
+
+// defaultBulkPublisher is the default implementation of BulkPublisher.
+// It is used when the component does not implement BulkPublisher.
+type defaultBulkPublisher struct {
+	p contribPubsub.PubSub
+}
+
+// NewDefaultBulkPublisher returns a new defaultBulkPublisher from a PubSub.
+func NewDefaultBulkPublisher(p contribPubsub.PubSub) *defaultBulkPublisher {
+	return &defaultBulkPublisher{
+		p: p,
+	}
+}
+
+// BulkPublish publishes a list of messages to a topic as individual Publish requests.
+// If 'bulkPublishSerially' metadata is set to true, the messages are sent to the broker serially,
+// in the same order. Otherwise they are sent to the broker as parallel Publish requests.
+func (p *defaultBulkPublisher) BulkPublish(_ context.Context, req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	if req.Metadata[bulkPublishSeriallyKey] == "true" {
+		return p.bulkPublishSerial(req)
+	} else {
+		return p.bulkPublishParallel(req)
+	}
+}
+
+// bulkPublishSerial publishes messages in a serial order. This is slower, but ensures
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishSerial(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+	hasError := false
+
+	for _, entry := range req.Entries {
+		status := p.bulkPublishSingleEntry(req.PubsubName, req.Topic, entry)
+		if status.Error != nil {
+			hasError = true
+		}
+
+		statuses = append(statuses, status)
+	}
+
+	var err error
+	if hasError {
+		err = ErrBulkPublishFailure
+	}
+
+	return contribPubsub.BulkPublishResponse{Statuses: statuses}, err
+}
+
+// bulkPublishParallel publishes messages in parallel. This is faster, but does not guarantee
+// that messages are sent to the broker in the same order as specified in the request.
+func (p *defaultBulkPublisher) bulkPublishParallel(req *contribPubsub.BulkPublishRequest) (contribPubsub.BulkPublishResponse, error) {
+	statuses := make([]contribPubsub.BulkPublishResponseEntry, 0, len(req.Entries))
+	var eg errgroup.Group
+	eg.SetLimit(utils.GetIntOrDefault(req.Metadata, bulkPublishMaxConcurrencyKey, defaultBulkPublishMaxConcurrency))
+
+	statusChan := make(chan contribPubsub.BulkPublishResponseEntry, len(req.Entries))
+
+	for _, entry := range req.Entries {
+		entry := entry
```

Comment:
Nit: this is kinda ugly :) How about:

```go
for i := range req.Entries {
		entry := req.Entries[i]
```

Doesn't change the substance but at least doesn't have a weird assignment to a variable with the same name

Reply:
Addressed via https://github.com/dapr/dapr/pull/5396